### PR TITLE
DOC-8204 -- Auto-purge on Channel Access Revoke

### PR DESCRIPTION
**Pull request by ibsoln**
https://issues.couchbase.com/browse/DOC-8204

### CODE REVIEW

File: modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc [hunk 1]
```diff
@@ -160,6 +174,148 @@ include::{root-partials}replication-filters.adoc[]
...
+
+* User loses direct access to channel
+* User is removed from a role
+* A role that user belongs to is revoked access to channel
+
+In 3.x, by default, when a user loses access to a channel, the next Couchbase Lite Pull replication auto-purges all documents in the channel from local Couchbase Lite databases (on devices belonging to the user) *unless* they belong to any of the user’s other channels -- see: <<tbl-revoke-behavior>>.
+
+Documents that exist in multiple channels belonging to the user (even if they are not actively replicating that channel) are not auto-purged unless the user loses access to all channels.
+
+Users will be receive an `AccessRemoved` notification from the DocumentListener if they lose document access due to channel access revocation; this is sent regardless of the current auto-purge setting.
```
Comment (JRascagneres) [inline modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc:199]: Needs the 'be' removing.

File: modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc [hunk 2]
```diff
@@ -160,6 +174,148 @@ include::{root-partials}replication-filters.adoc[]
...
+Clients wanting to control whether to sync previous auto purged versions of documents (rather than pull down purged documents) must also move the documents out of all of the users' channels so they are not synced down again.
```
Comment (JRascagneres) [inline modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc:200]: Not sure I follow this sentence? 
What are we trying to say here?

File: modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc [hunk 3]
```diff
@@ -160,6 +174,148 @@ include::{root-partials}replication-filters.adoc[]
...
+^|Impact on Sync
+
+.>h|Replication Type
+^.>h|Access Control on Sync Gateway
+^.>h|Expected behavior when _enable_auto_purge=true_
+
+|Pull only
+|User revoked access to channel.
+
+Sync Function includes `requireAccess(revokedChannel)`
```
Comment (JRascagneres) [inline modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc:216]: Just the fact that the user lost access to the channel will result in a revocation. `requireAccess` is not required for pull revocation.

File: modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc (hunk 2 above)
Comment (ibsoln) [reply modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc:200]: It's an all but straight lift from the PRD; which could allegedly be code for 'I'm not sure either'. I'll find out more and resolve.

File: modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc (hunk 1 above)
Comment (ibsoln) [reply modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc:199]: ... if you knew how long it took me to find that 'be' 

removed

File: modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc (hunk 3 above)
Comment (ibsoln) [reply modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc:216]: removed requireaccess

File: modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc [hunk 4]
```diff
@@ -160,6 +174,148 @@ include::{root-partials}replication-filters.adoc[]
...
 
+
+[#anchor-auto-purge-on-revoke]
+=== Auto-purge on Channel Access Revocation
+
+.Breaking Change
+[CAUTION]
+--
+By default, when a user loses access to a channel all documents in the channel (that do not also belong to any of user’s other channels) are auto-purged from the local database (in devices belonging to the user). +
+_In release 2.x these documents will remain in the local database._
```
Comment (pasin) [inline modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc:185]: I'm not sure if trying to explain it here is too complicated.

In 2.x, CBL also does auto-purge but only for the case when the user loses access to the doc via removing the doc from all of the channels belong to the user. (Not sure if we have this explained any where in 2.x doc).

In 3.0, in addition to 2.x auto purge:
*  CBL will auto-purge the docs when the user loses access to the doc via channel access revocation.
* CBL provides configuration to to enable / disable auto purge - enable by default.

File: modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc [hunk 5]
```diff
@@ -160,6 +174,148 @@ include::{root-partials}replication-filters.adoc[]
...
+--
+
+==== Behavior
+Users may lose access to channels in a number of ways:
+
+* User loses direct access to channel
+* User is removed from a role
+* A role that user belongs to is revoked access to channel
```
Comment (pasin) [inline modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc:193]: What does the third bullet mean?

File: modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc (hunk 1 above)
Comment (pasin) [reply modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc:199]: It is DocumentReplicationListener not DocumentListener.

File: modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc (hunk 2 above)
Comment (pasin) [reply modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc:200]: I don't understand this either.

File: modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc [hunk 6]
```diff
@@ -160,6 +174,148 @@ include::{root-partials}replication-filters.adoc[]
...
+|User revoked access to channel +
+Sync Function includes `requireAccess(revokedChannel)`
+|Previously synced documents are auto purged on Couchbase Lite.
+
+Local changes continue to be  pushed to remote but are rejected by Sync Gateway
+
+|===
+
+
+If a user subsequently regains access to a lost channel, then any previously auto-purged documents still assigned to any of their channels are automatically pulled down by the active Sync Gateway -- see behavior summary in <<tbl-regain-behavior>>
```
Comment (pasin) [inline modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc:235]: The document will be synced back only when there is new update.

File: modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc [hunk 7]
```diff
@@ -160,6 +174,148 @@ include::{root-partials}replication-filters.adoc[]
...
+2+|System State
+^|Impact on Sync
+
+.>h|Replication Type
+^.>h|Access Control on Sync Gateway
+^.>h|Expected behavior when _enable_auto_purge=true_
+
+|Pull only
+|User REASSIGNED access to channel
+|Previously purged documents that are still in the channel are automatically pulled by couchbase lite
```
Comment (pasin) [inline modules/ROOT/pages/_partials/commons/common-sgw-replication.adoc:250]: Only when there is new update to the docs.

File: modules/android/examples/snippets/app/src/main/java/com/couchbase/code_snippets/Examples.java [hunk 8]
```diff
@@ -1142,7 +1142,17 @@ public void testCustomRetryConfig() throws URISyntaxException {
...
 
+    public void docsSetAutoPurge() throws CouchbaseliteException {
 
+      DatabaseConfiguration config = new DatabaseConfiguration();
+      Database database1 = new Database("mydb", config);
+
+      ReplicatorConfiguration repcfg =
+      new ReplicatorConfiguration(database, target);
+
+      repcfg.setAutoPurgeEnabled(true) // <.>
```
Comment (pasin) [inline modules/android/examples/snippets/app/src/main/java/com/couchbase/code_snippets/Examples.java:1153]: Missing semi-colon.

File: modules/swift/examples/code_snippets/SampleCodeTest.swift [hunk 9]
```diff
@@ -3719,6 +3719,11 @@ class myActPeerClass {
     config.replicatorType = .pushAndPull
 
     // end::p2p-act-rep-config-type[]
+    // tag::autopurge-override[]
+    // set auto-purge behavior (here we override default)
+    thisConfig.enableAutoPurge = false; // <.>
```
Comment (pasin) [inline modules/swift/examples/code_snippets/SampleCodeTest.swift:3724]: No semi colon

**Pull request merged**